### PR TITLE
refactor: decouple DiscoLog from global configuration

### PR DESCRIPTION
**Pull request by martosaur**
### Contributor checklist
- [x] My commit messages follow the [Conventional Commit Message Format](https://gist.github.com/stephenparish/9941e89d80e2bc58a153#format-of-the-commit-message)
- [x] Features include unit/acceptance tests

Context: https://elixirforum.com/t/discolog-use-discord-as-a-logging-service-and-error-tracking-solution/65730/11

As per our discussion, here's my attempt on decoupling DiscoLog from the global application configuration. I'm not proud of the size of this PR, but the configuration is indeed _global_ and touches pretty much everything. I recommend reviewing modules in the same order as I describe the changes here.

###  `DiscoLog.Config`

This module is now the place responsible for DiscoLog configuration.

`Config.validate/1` and `Config.validate!/1` are helper functions that validate raw keyword list/map with the help of NimbleOptions.

`Config.read!/0` is the only place in the app where we read from the application configuration. It is meant to be used whenever we consciously decided that we want to use options defined under `config :disco_log, ...`. For example, when starting the application. This function has a special case: if we disabled logging all together with `config :disco_log, enable: false`, it will simply return `%{enable: false}`. This simply allows users to disable `disco_log` and manage everything themselves.

There is also a new option here, `supervisor_name`, which defines which supervisor tree should be used. More on the supervisor below.

### `DiscoLog.Application`
Changes here:
1. First of all, on a startup we read global configuration with `Config.read!/0`.
2. If logging is enabled, the app starts all necessary processes and attaches handlers, _passing config directly to them_.

### `DiscoLog.Supervisor`
This is a new module that's responsible for starting all necessary workers. It accepts a validated configuration and:
1. Registers itself under name specified in `config.supervisor_name`
2. Starts a registry process that will be used by its workers
3. Starts `DiscoLog.Storage` worker
4. Starts `DiscoLog.Dedupe` worker

Users can start as many supervisors as they want, they just need to have different names. The whole setup is very similar to how Oban does things.

### `DiscoLog.Dedupe`
Changes:
1. This worker now registers itself in its supervisor's registry instead of global namespace.
2. Instead of using named ets table, it stores table identifier in its state _as well as in the registry_, so that anyone could access it without making a call to the worker.
3. `Dedupe.insert/2` now requires a supervisor name as a first argument to identify which dedupe to use. With this information it can fetch ets table from supervisor's registry and do what needs to be done.

### `DiscoLog.Storage`
1. In the same manner as `Dedupe`, `Storage` registers itself in the registry.
2. _It also directly uses registry's key value store to store occurrence threads_. This ends up to be stored in the ets anyway, just managed by Registry itself. 
3. Client functions changed to accept supervisor name as first argument, since storage isn't a globally registered singleton anymore.

### `DiscoLog.Registry`
A new module with two helper functions. Basically, it encapsulates supervisor tree naming policy.

### `DiscoLog.Discord.Config`
Now this is an awkward place. Discord config has a sizable overlap with DiscoLog config. I feel like there is a room for improvement here, but it wasn't necessary for the purpose of this PR, so I made Discord config a struct which can be fully composed from DiscoLog config and is stored right there under `discord_config` key.

### `DiscoLog.Discord` 
This module is a congregation of `DiscordBehaviour`, `Discord` and `DiscordImpl`. The purpose of `Discord` was to swap between `DiscordImpl` and `DiscordMock` based on global configuration. This isn't necessary anymore as discord implementation module is set directly in the config, so `Discord` was removed and `DiscordImpl` was renamed into `Discord`. 

Another major change here is that all `Discord` functions accept `config.discord_config` as first argument. This unlocks usage of multiple bots/servers/etc.

### `DiscoLog.LoggerHandler`
1. `LoggerHandler` does not have its own config anymore, it uses the same configuration produced by `DiscoLog.Config.validate`.
2. It now passes this config to downstream functions from `DiscoLog.Client` module.
3. The same applies to `Error.new/3` (which is now `Error.new/4`), but it also requires an `otp_app`, not the full config, so we pass only that.

### `DiscoLog.Discord.Client`, `DiscoLog.Discord.Context`, `DiscoLog.Client`
A lot of fairly simple refactoring here to pass `config` map around explicitly.

### `DiscoLog`

`DiscoLog.report/3` function now accepts an optional `config`. Otherwise, it will use global configuration with `Config.read!`

A side note: to be quite honest, I don't fully understand why this function exists. I feel like it might be inspired by `Sentry.capture_exception`, but would anyone want to report an error directly to Sentry without logging it? 🤔 

### Tests

Previously, tests were synchronous and for the most part relied on `before_send` callback for assertions. I went ahead and reworked tests to be run mostly async and replaced `before_send` with Mox mocks.

Basically, tests now start processes they need, attach handlers and are generally independent from each other. The only place where they interfere is through logger handlers, which are global and may be called on events that originate from other tests. I worked around that by attaching a logger filter to the handler which piggybacks on Mox ownership mechanism and drops all events that aren't coming from the test.

The only 2 tests that are synchronous are
* the one where we restart logger application to enable sasl reports
* and one where we check that by default `:disco_log` application will boot supervisor on startup.

Ultimately, with this PR users should be able to mix and match handlers, bots, servers, etc. however they want, while the existing default way should still work.

There's a lot to add in terms of documentation etc., but I figured it's better to start the conversation and agree on direction and do final polishing later 😅

### CODE REVIEW

**Review by mrdotb — CHANGES_REQUESTED** [review]
:wave: 
It's a lot of changes so I may have missed stuff.
The test rework are really good ! I was not very satisfied off the curent way of doing it.

I tested a bit with the `dev.exs` everything seems to works find with the current config.

For the changes.
The req client part is a bit hard to follow I put some reviews about that.

For the documentation the existing one is still valid and a good default since you could make the modification without breaking change.
We need one for running disco_log in your own supervision tree.
And something for a more fancy setup

File: lib/disco_log/dedupe.ex [hunk 1]
```diff
@@ -9,54 +9,79 @@ defmodule DiscoLog.Dedupe do
...
+          _inserted_new? = :ets.insert_new(table, {hash, now}) -> :new
+          true -> :existing
+        end
 
-    cond do
-      _found? = :ets.update_element(@ets, hash, {_position = 2, now}) -> :existing
-      _inserted_new? = :ets.insert_new(@ets, {hash, now}) -> :new
-      true -> :existing
+      _ ->
+        :new
```
[the remark this answers is not in the document]
Comment (martosaur) [reply lib/disco_log/dedupe.ex:44]: My reasoning was that there is always a chance the process won't be there. On a start up, we attach handler before supervision tree is started, so any logs emitted in this window won't be able to resolve `Dedupe` process. Although, the same can be said about Registry process too which we use for a lookup 🤔 Yeah, I think it's better to be more assertive here and if that's ever a problem we can either make handler catch exceptions or something similar 👍

File: lib/disco_log/discord/client.ex [hunk 2]
```diff
@@ -127,15 +118,15 @@ defmodule DiscoLog.Discord.Client do
     end
   end
 
-  defp client do
+  def client(token, log? \\ false) do
```
[the remark this answers is not in the document]
Comment (martosaur) [reply lib/disco_log/discord/client.ex:121]: Yeah I can do this! I have my thoughts on discord/client/context family of modules, but that can be totally discussed in a separate issue/PR 👍

File: lib/disco_log/discord/context.ex [hunk 3]
```diff
@@ -6,22 +6,26 @@ defmodule DiscoLog.Discord.Context do
...
       channel when is_map(channel) ->
         {:ok, channel}
 
       nil ->
-        channel_config
-        |> maybe_add_parent_id(parent_id)
-        |> Discord.Client.create_channel()
+        params = maybe_add_parent_id(channel_config, parent_id)
+
+        config.token
```
[the remark this answers is not in the document]
Comment (martosaur) [reply lib/disco_log/discord/context.ex:17]: I can see the appeal. I went ahead and also changed other places where `guild_id` or `occurrence_channel_id` were passed as parameters but could be fetched from the config

File: lib/disco_log/storage.ex [hunk 4]
```diff
@@ -4,54 +4,71 @@ defmodule DiscoLog.Storage do
...
+    {:ok, state, {:continue, :restore}}
   end
 
   @impl GenServer
   def handle_continue(:restore, state) do
-    Discord.list_occurrence_threads()
-    |> Enum.each(fn {fingerprint, thread_id} ->
-      :ets.insert(@ets, {fingerprint, thread_id})
+    Registry.update_value(state.registry, __MODULE__, fn _ ->
+      state.discord_config
```
[the remark this answers is not in the document]
Comment (martosaur) [reply lib/disco_log/storage.ex:58]: Hmm, you mean you'd rather match on state in the function head and have just `module.list_occurrence_threads(...`? Another option I can think of is using `apply`, but not sure if it's better.

**Comment by martosaur** [comment]
Thanks for the quick review! I pushed some changes. Please take another look and if things look reasonable I'll start working towards writing docs.

**Comment by mrdotb** [comment]
Looks good to me with those changes.

**Comment by martosaur** [comment]
Added docs!

**Review by mrdotb — COMMENTED** [review]
Looks good but the example is maybe too simple ? 🤔
How could I use two disco_log in my application ?

**Comment by martosaur** [comment]
Changed the example to showcase logging to 2 servers! Also pushed a small QoL improvement to Supervisor to make this scenario simpler

**Comment by mrdotb** [comment]
:wave: Sorry for the delay. 
Good to merge !
One last thing could you squash all your commits.
I will do a release this week end.

**Comment by martosaur** [comment]
Done!
